proxy/src/telemetry/metrics/prometheus.rs

```diff
@@ -0,0 +1,521 @@
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.inner.into_iter()
+    }
+}
+
+impl Labels {
+
+    pub fn with_proxy_ctx<'a>(&self, proxy_ctx: &'a Arc<ctx::Proxy>) -> Labels {
+        let direction = if proxy_ctx.is_inbound() {
+            "inbound"
+        } else {
+            "outbound"
+        };
+        let inner = self.child() + ("direction", direction.to_owned());
+        let inner = Arc::new(inner);
+        Labels {
+            inner,
+        }
+    }
+
+    pub fn with_request_ctx<'a>(&self, req: &'a Arc<ctx::http::Request>)
+        -> Labels
+    {
+        let authority = req.uri
+            .authority_part()
+            .map(http::uri::Authority::to_string)
+            .unwrap_or_else(String::new);
+        let inner = self.child() + ("authority", authority);
+        let inner = Arc::new(inner);
+        Labels {
+            inner,
+        }
+    }
+
+    pub fn with_response_ctx<'a>(&self, rsp: &'a Arc<ctx::http::Response>)
+        -> Labels
+    {
+        let status = rsp.status.as_str();
+        let authority = rsp.request.uri
+            .authority_part()
+            .map(http::uri::Authority::to_string)
+            .unwrap_or_else(String::new);
+        let inner = self.child() + ("authority", authority) +
+            ("status_code", status.into());
+        let inner = Arc::new(inner);
+        Labels {
+            inner,
+        }
+    }
+
+
+    pub fn with_grpc_status(&self, status: Option<u32>)
+        -> Labels {
+        let inner = if let Some(code) = status {
+            Arc::new(self.child() + ("grpc_status", format!("{}", code)))
+        } else {
+            self.inner.clone()
+        };
+        Labels {
+            inner,
+        }
+    }
+
+    fn child(&self) -> LabelsInner {
+        LabelsInner {
+            parent: Some(self.clone()),
+            ..Default::default()
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+}
+
+impl fmt::Display for Labels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut labels = self.into_iter();
+
+        if let Some((label, value)) = labels.next() {
+            // format the first label pair without a comma.
+            write!(f, "{}=\"{}\"", label, value)?;
+
+            // format the remaining pairs with a comma preceeding them.
+            for (label, value) in labels {
+                write!(f, ",{}=\"{}\"", label, value)?;
+            }
+        }
+
+        Ok(())
+    }
+
+}
+
+// ===== impl LabelsInner =====
+
+impl LabelsInner {
+
+    fn new() -> Self {
+        Default::default()
+    }
+
+    fn len(&self) -> usize {
+        self.labels.len() + self.parent
+            .as_ref()
+            .map(Labels::len)
+            .unwrap_or(0)
+    }
+
+    fn parent_iter<'a>(&'a self)
+        -> Box<Iterator<Item=(&'a str, &'a str)> + 'a>
```

Comment:
Would it be super hard to remove this `Box`? Maybe instead of returning iterators, labels could just be passed a closure to execute for each?